client/src/components/Minimap/Minimap.tsx

```diff
@@ -0,0 +1,238 @@
+  anchorsContainer: HTMLElement;
+  links: readonly MinimapMenuItem[];
+  onUpdate: (container: HTMLElement) => void;
+  toggleAllPanels: (expanded: boolean) => void;
+}
+
+const observerOptions = {
+  root: null,
+  // Count an element as "in", accounting for the 50px slim header and 70px actions footer.
+  rootMargin: '-50px 0px -70px 0px',
+  // 10% visibility within the boxed viewport is enough.
+  threshold: 0.1,
+};
+
+type LinkIntersections = {
+  [href: string]: boolean;
+};
+
+const mapIntersections = (
+  acc: LinkIntersections,
+  { target, isIntersecting }: IntersectionObserverEntry,
+) => {
+  const href = `#${target.closest('[data-panel]')?.id}` || '';
+  acc[href] = isIntersecting;
+  return acc;
+};
+
+/**
+ * For cases where the minimap has more item than can fit in the viewport,
+ * we need to keep its scroll position updated to follow page scrolling.
+ */
+const updateScrollPosition = (list: HTMLOListElement) => {
+  const activeLinks = list.querySelectorAll<HTMLElement>(
+    'a[aria-current="true"]',
+  );
+
+  if (activeLinks.length === 0) {
+    return;
+  }
+
+  const firstActive = activeLinks[0];
+  const lastActive = activeLinks[activeLinks.length - 1];
+  let newScroll = list.scrollTop;
+  if (firstActive) {
+    if (firstActive.offsetTop < list.scrollTop) {
+      newScroll = firstActive.offsetTop;
+    }
+  }
+  if (lastActive) {
+    if (lastActive.offsetTop > list.scrollTop + list.offsetHeight) {
+      newScroll =
+        lastActive.offsetTop - list.offsetHeight + lastActive.offsetHeight;
+    }
+  }
+
+  // Scroll changes require mutating this property.
+  // eslint-disable-next-line no-param-reassign
+  list.scrollTop = newScroll;
+};
+
+/**
+ * Minimap sidebar menu, with one internal link per section of the page.
+ * The minimap has a lot of advanced behavior:
+ * - It opens and closes based on hover, except if interacted with.
+ * - It also opens and closes when clicking its toggle.
+ * - It closes when clicking outside.
+ * - It uses IntersectionObserver to display which menu items are currently "visible" on the page.
+ */
+const Minimap: React.FunctionComponent<MinimapProps> = ({
+  container,
+  anchorsContainer,
+  links,
+  onUpdate,
+  toggleAllPanels,
+}) => {
+  const [expanded, setExpanded] = useState<boolean>(false);
+  // Keep track of whether we should keep the minimap expanded (should auto-close on mouseout if not interacted with).
+  const [keepExpanded, setKeepExpanded] = useState<boolean>(false);
+  // Collapse all yes/no state.
+  const [panelsExpanded, setPanelsExpanded] = useState<boolean>(true);
+  const [intersections, setIntersections] = useState<LinkIntersections>({});
+  const observer = useRef<IntersectionObserver | null>(null);
+  const lastIntersections = useRef({});
+  const updateLinks = useRef<CallableFunction | null>(null);
+  const listRef = useRef<HTMLOListElement>(null);
+
+  // Keep track of all the different ways the minimap can be opened and closed.
```

Comment:
All of these aren’t so complex but there’s a fair few to keep track of, so I put them all one after the other.